File: tracker/src/Ironmon.Tracker.Protocol/Lookup/GenerationProfile.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace Ironmon.Tracker.Protocol.Lookup;

/// <summary>
/// Defines the complete set of deterministic algorithm families pinned by a generation profile.
/// </summary>
public static class GenerationAlgorithmNames
{
    /// <summary>
    /// Gets the deterministic hash-contract family name.
    /// </summary>
    public const string HashContract = "hash_contract";

    /// <summary>
    /// Gets the custom-fusion eligibility family name.
    /// </summary>
    public const string CustomFusionEligibility = "custom_fusion_eligibility";

    /// <summary>
    /// Gets the species-mapping family name.
    /// </summary>
    public const string SpeciesMapping = "species_mapping";

    /// <summary>
    /// Gets the ability-assignment family name.
    /// </summary>
    public const string AbilityAssignment = "ability_assignment";

    /// <summary>
    /// Gets the base-stat generation family name.
    /// </summary>
    public const string BaseStats = "base_stats";

    /// <summary>
    /// Gets the move-access generation family name.
    /// </summary>
    public const string MoveAccess = "move_access";

    /// <summary>
    /// Gets the item-slot generation family name.
    /// </summary>
    public const string ItemSlots = "item_slots";

    /// <summary>
    /// Gets the normal-evolution generation family name.
    /// </summary>
    public const string NormalEvolution = "normal_evolution";

    /// <summary>
    /// Gets the fusion-evolution generation family name.
    /// </summary>
    public const string FusionEvolution = "fusion_evolution";

    /// <summary>
    /// Gets the player fusion and reversal family name.
    /// </summary>
    public const string PlayerFusionReversal = "player_fusion_reversal";

    /// <summary>
    /// Gets the gym-party expansion family name.
    /// </summary>
    public const string GymPartyExpansion = "gym_party_expansion";

    /// <summary>
    /// Gets the caught-fusion component family name.
    /// </summary>
    public const string CaughtFusionComponent = "caught_fusion_component";

    /// <summary>
    /// Gets the progression-support Pokémon family name.
    /// </summary>
    public const string ProgressionSupportPokemon = "progression_support_pokemon";

    /// <summary>
    /// Gets every required family name in ordinal order.
    /// </summary>
    public static IReadOnlyList<string> All { get; } =
    [
        AbilityAssignment,
        BaseStats,
        CaughtFusionComponent,
        CustomFusionEligibility,
        FusionEvolution,
        GymPartyExpansion,
        HashContract,
        ItemSlots,
        MoveAccess,
        NormalEvolution,
        PlayerFusionReversal,
        ProgressionSupportPokemon,
        SpeciesMapping
    ];
}

/// <summary>
/// Identifies one immutable implementation of a deterministic algorithm family.
/// </summary>
public sealed class GenerationAlgorithmPayload
{
    /// <summary>
    /// Initializes an empty algorithm descriptor for protocol serialization.
    /// </summary>
    public GenerationAlgorithmPayload()
    {
    }

    /// <summary>
    /// Gets or initializes the stable algorithm-family name.
    /// </summary>
    public required string Name { get; init; }

    /// <summary>
    /// Gets or initializes the family-specific implementation version.
    /// </summary>
    public int Version { get; init; }
}

/// <summary>
/// Identifies one immutable data component used by deterministic generation.
/// </summary>
public sealed class GenerationComponentPayload
{
    /// <summary>
    /// Initializes an empty component descriptor for protocol serialization.
    /// </summary>
    public GenerationComponentPayload()
    {
    }

    /// <summary>
    /// Gets or initializes the stable component name.
    /// </summary>
    public required string Name { get; init; }

    /// <summary>
    /// Gets or initializes the component's independently versioned storage schema.
    /// </summary>
    public int SchemaVersion { get; init; }

    /// <summary>
    /// Gets or initializes the lowercase SHA-256 digest of the exact component bytes.
    /// </summary>
    public required string Sha256 { get; init; }

    /// <summary>
    /// Gets or initializes the exact component byte length.
    /// </summary>
    public long ByteLength { get; init; }
}

/// <summary>
/// Describes the immutable algorithms and data snapshot needed to generate or inspect one run.
/// </summary>
public sealed class GenerationProfilePayload
{
    /// <summary>
    /// Initializes an empty generation profile for protocol serialization.
    /// </summary>
    public GenerationProfilePayload()
    {
    }

    /// <summary>
    /// Gets or initializes the generation-profile contract schema.
    /// </summary>
    public int SchemaVersion { get; init; } = GenerationProfileFingerprint.SchemaVersion;

    /// <summary>
    /// Gets or initializes every deterministic algorithm descriptor.
    /// </summary>
    public IReadOnlyList<GenerationAlgorithmPayload> Algorithms { get; init; } = [];

    /// <summary>
    /// Gets or initializes every immutable data-component descriptor.
    /// </summary>
    public IReadOnlyList<GenerationComponentPayload> Components { get; init; } = [];
}

/// <summary>
/// Validates, canonicalizes, and fingerprints immutable generation profiles.
/// </summary>
public static class GenerationProfileFingerprint
{
    private const string LowerHexCharacters = "0123456789abcdef";

    /// <summary>
    /// Gets the only generation-profile schema currently supported.
    /// </summary>
    public const int SchemaVersion = 1;

    /// <summary>
    /// Creates the lowercase SHA-256 identity of a validated generation profile.
    /// </summary>
    /// <param name="profile">The profile to validate and fingerprint.</param>
    /// <returns>The lowercase SHA-256 profile identity.</returns>
    public static string Create(GenerationProfilePayload profile)
    {
        byte[] canonicalJson = CreateCanonicalJson(profile);
        return Convert.ToHexString(SHA256.HashData(canonicalJson)).ToLowerInvariant();
    }

    /// <summary>
    /// Creates the canonical UTF-8 JSON representation used for profile identity.
    /// </summary>
    /// <param name="profile">The profile to validate and canonicalize.</param>
    /// <returns>The canonical UTF-8 JSON bytes.</returns>
    public static byte[] CreateCanonicalJson(GenerationProfilePayload profile)
    {
        GenerationProfilePayload normalized = Normalize(profile);
        JsonElement document = JsonSerializer.SerializeToElement(normalized, TrackerJson.Options);
        using MemoryStream stream = new();
        using (Utf8JsonWriter writer = new(stream))
            WriteCanonical(writer, document);

        return stream.ToArray();
    }

    /// <summary>
    /// Validates a profile and returns its deterministically ordered representation.
    /// </summary>
    /// <param name="profile">The profile to validate.</param>
    /// <returns>A normalized profile.</returns>
    private static GenerationProfilePayload Normalize(GenerationProfilePayload profile)
    {
        ArgumentNullException.ThrowIfNull(profile);
        if (profile.SchemaVersion != SchemaVersion)
            throw new ArgumentException($"Unsupported generation profile schema {profile.SchemaVersion}.", nameof(profile));

        if (profile.Algorithms.Count != GenerationAlgorithmNames.All.Count)
            throw new ArgumentException("The generation profile does not contain every required algorithm family.", nameof(profile));

        GenerationAlgorithmPayload[] algorithms = [.. profile.Algorithms.OrderBy(entry => entry.Name, StringComparer.Ordinal)];
        for (int index = 0; index < algorithms.Length; index++)
        {
            GenerationAlgorithmPayload algorithm = algorithms[index];
            if (!IsValidName(algorithm.Name) || algorithm.Version <= 0)
                throw new ArgumentException("The generation profile contains an invalid algorithm descriptor.", nameof(profile));

            if (!StringComparer.Ordinal.Equals(algorithm.Name, GenerationAlgorithmNames.All[index]))
                throw new ArgumentException("The generation profile does not contain exactly the required algorithm families.", nameof(profile));
        }

        if (profile.Components.Count == 0)
            throw new ArgumentException("The generation profile must contain at least one data component.", nameof(profile));

        GenerationComponentPayload[] components = [.. profile.Components.OrderBy(entry => entry.Name, StringComparer.Ordinal)];
        for (int index = 0; index < components.Length; index++)
        {
            GenerationComponentPayload component = components[index];
            if (!IsValidName(component.Name) || component.SchemaVersion <= 0 || component.ByteLength <= 0 || !IsLowerSha256(component.Sha256))
                throw new ArgumentException("The generation profile contains an invalid data-component descriptor.", nameof(profile));

            if (index > 0 && StringComparer.Ordinal.Equals(component.Name, components[index - 1].Name))
                throw new ArgumentException("Generation component names must be unique.", nameof(profile));
        }

        return new GenerationProfilePayload
        {
            SchemaVersion = SchemaVersion,
            Algorithms = algorithms,
            Components = components
        };
    }

    /// <summary>
    /// Determines whether a contract name uses the portable lowercase identifier grammar.
    /// </summary>
    /// <param name="value">The value to inspect.</param>
    /// <returns>Whether the name is valid.</returns>
    private static bool IsValidName(string? value)
    {
        if (string.IsNullOrEmpty(value) || value[0] is < 'a' or > 'z')
            return false;

        return value.All(character => character is >= 'a' and <= 'z' or >= '0' and <= '9' or '_');
    }

    /// <summary>
    /// Determines whether a value is an exact lowercase SHA-256 digest.
    /// </summary>
    /// <param name="value">The value to inspect.</param>
    /// <returns>Whether the digest is valid.</returns>
    private static bool IsLowerSha256(string? value)
    {
        return value is { Length: 64 } && value.All(LowerHexCharacters.Contains);
    }

    /// <summary>
    /// Writes JSON with object properties sorted ordinally at every depth.
    /// </summary>
    /// <param name="writer">The destination JSON writer.</param>
    /// <param name="element">The JSON value to write.</param>
    private static void WriteCanonical(Utf8JsonWriter writer, JsonElement element)
    {
        if(element.ValueKind == JsonValueKind.Object)
        {
            writer.WriteStartObject();
            foreach (JsonProperty property in element.EnumerateObject().OrderBy(property => property.Name, StringComparer.Ordinal))
            {
                writer.WritePropertyName(property.Name);
                WriteCanonical(writer, property.Value);
            }

            writer.WriteEndObject();
            return;
        }

        if(element.ValueKind == JsonValueKind.Array)
        {
            writer.WriteStartArray();
            foreach (JsonElement item in element.EnumerateArray())
                WriteCanonical(writer, item);

            writer.WriteEndArray();
            return;
        }

        if(element.ValueKind == JsonValueKind.String)
        {
            writer.WriteStringValue(element.GetString());
            return;
        }

        if(element.ValueKind == JsonValueKind.Number)
        {
            writer.WriteRawValue(element.GetRawText());
            return;
        }

        if(element.ValueKind == JsonValueKind.True)
        {
            writer.WriteBooleanValue(true);
            return;
        }

        if (element.ValueKind == JsonValueKind.False)
        {
            writer.WriteBooleanValue(false);
            return;
        }

        if (element.ValueKind == JsonValueKind.Null)
        {
            writer.WriteNullValue();
            return;
        }

        throw new ArgumentException("The generation profile contains unsupported JSON data.", nameof(element));
    }
}

File: src/compatibility/Generation_Profile.rb
require "digest/sha2"

module Ironmon
  module GenerationProfile
    SCHEMA_VERSION = 1
    SHA256_PATTERN = /\A[0-9a-f]{64}\z/
    NAME_PATTERN = /\A[a-z][a-z0-9_]*\z/
    ALGORITHM_FAMILIES = [
      "hash_contract",
      "custom_fusion_eligibility",
      "species_mapping",
      "ability_assignment",
      "base_stats",
      "move_access",
      "item_slots",
      "normal_evolution",
      "fusion_evolution",
      "player_fusion_reversal",
      "gym_party_expansion",
      "caught_fusion_component",
      "progression_support_pokemon"
    ].sort.freeze
    CURRENT_ALGORITHM_VERSIONS = ALGORITHM_FAMILIES.each_with_object({}) do |name, result|
      result[name] = 1
    end.freeze

    def self.build(algorithms, components)
      manifest = {
        "schema_version" => SCHEMA_VERSION,
        "algorithms" => algorithms.map do |name, version|
          { "name" => name.to_s, "version" => version }
        end,
        "components" => components
      }
      return normalize(manifest)
    end

    def self.component(name, schema_version, path)
      bytes = File.binread(path)
      return {
        "name" => name.to_s,
        "schema_version" => schema_version,
        "sha256" => Digest::SHA256.hexdigest(bytes),
        "byte_length" => bytes.bytesize
      }
    end

    def self.fingerprint(manifest)
      return Digest::SHA256.hexdigest(canonical_json(normalize(manifest)))
    end

    def self.canonical_json(value)
      return generate_json(canonical_value(value))
    end

    def self.normalize(manifest)
      raise ArgumentError, "generation profile must be a hash" if !manifest.is_a?(Hash)
      schema_version = value_for(manifest, "schema_version")
      if schema_version != SCHEMA_VERSION
        raise ArgumentError, "unsupported generation profile schema #{schema_version.inspect}"
      end
      algorithms = normalize_algorithms(value_for(manifest, "algorithms"))
      components = normalize_components(value_for(manifest, "components"))
      return {
        "schema_version" => SCHEMA_VERSION,
        "algorithms" => algorithms,
        "components" => components
      }
    end

    def self.normalize_algorithms(algorithms)
      raise ArgumentError, "generation profile algorithms must be an array" if !algorithms.is_a?(Array)
      result = algorithms.map do |entry|
        raise ArgumentError, "generation algorithm must be a hash" if !entry.is_a?(Hash)
        name = value_for(entry, "name").to_s
        version = value_for(entry, "version")
        raise ArgumentError, "invalid generation algorithm name #{name.inspect}" if !NAME_PATTERN.match?(name)
        raise ArgumentError, "invalid generation algorithm version for #{name}" if !version.is_a?(Integer) || version <= 0
        { "name" => name, "version" => version }
      end.sort_by { |entry| entry["name"] }
      names = result.map { |entry| entry["name"] }
      raise ArgumentError, "generation algorithm names must be unique" if names.uniq.length != names.length
      if names != ALGORITHM_FAMILIES
        raise ArgumentError, "generation profile does not contain the required algorithm families"
      end
      return result
    end

    def self.normalize_components(components)
      raise ArgumentError, "generation profile components must be a non-empty array" if !components.is_a?(Array) || components.empty?
      result = components.map do |entry|
        raise ArgumentError, "generation component must be a hash" if !entry.is_a?(Hash)
        name = value_for(entry, "name").to_s
        schema_version = value_for(entry, "schema_version")
        sha256 = value_for(entry, "sha256").to_s
        byte_length = value_for(entry, "byte_length")
        raise ArgumentError, "invalid generation component name #{name.inspect}" if !NAME_PATTERN.match?(name)
        raise ArgumentError, "invalid component schema for #{name}" if !schema_version.is_a?(Integer) || schema_version <= 0
        raise ArgumentError, "invalid component digest for #{name}" if !SHA256_PATTERN.match?(sha256)
        raise ArgumentError, "invalid component byte length for #{name}" if !byte_length.is_a?(Integer) || byte_length <= 0
        {
          "name" => name,
          "schema_version" => schema_version,
          "sha256" => sha256,
          "byte_length" => byte_length
        }
      end.sort_by { |entry| entry["name"] }
      names = result.map { |entry| entry["name"] }
      raise ArgumentError, "generation component names must be unique" if names.uniq.length != names.length
      return result
    end

    def self.canonical_value(value)
      case value
      when Hash
        result = {}
        value.keys.sort_by(&:to_s).each do |original_key|
          key = original_key.to_s
          if result.key?(key)
            raise ArgumentError, "canonical JSON object keys must be unique after string conversion"
          end
          result[key] = canonical_value(value[original_key])
        end
        return result
      when Array
        return value.map { |entry| canonical_value(entry) }
      when Symbol
        return value.to_s
      when String, Integer, Float, TrueClass, FalseClass, NilClass
        return value
      end
      raise ArgumentError, "unsupported canonical JSON value #{value.class}"
    end

    def self.generate_json(value)
      case value
      when Hash
        entries = value.map do |key, child|
          "#{json_string(key)}:#{generate_json(child)}"
        end
        return "{#{entries.join(",")}}"
      when Array
        return "[#{value.map { |child| generate_json(child) }.join(",")}]"
      when String
        return json_string(value)
      when TrueClass, FalseClass
        return value.to_s
      when NilClass
        return "null"
      when Numeric
        raise ArgumentError, "non-finite numbers are not valid JSON" if value.respond_to?(:finite?) && !value.finite?
        return value.to_s
      end
      raise ArgumentError, "unsupported canonical JSON value #{value.class}"
    end

    def self.json_string(value)
      escaped = value.to_s.each_codepoint.map do |codepoint|
        case codepoint
        when 0x08 then "\\b"
        when 0x09 then "\\t"
        when 0x0A then "\\n"
        when 0x0C then "\\f"
        when 0x0D then "\\r"
        when 0x22 then '\\"'
        when 0x5C then "\\\\"
        else
          codepoint < 0x20 ? sprintf("\\u%04x", codepoint) :
            codepoint.chr(Encoding::UTF_8)
        end
      end.join
      return "\"#{escaped}\""
    end

    def self.value_for(hash, key)
      return hash[key] if hash.key?(key)
      symbol = key.to_sym
      return hash[symbol] if hash.key?(symbol)
      return nil
    end
  end
end
